README.md
# facebook_posts

A Flutter Package that queries posts from a facebook page to be used in your Flutter app. This is done without the aid of Facebook API, so no worrying of registration and limitations.

Credits to ayob#4578 (Discord) for writing the core logic of the package.


## Using

You can either wrap the method callback inside a future, to call the method. Or build up a custom Widget Tree to get the Posts.

The method **getFbData()** returns a list of **"FbPost"** Object. The Object has two properties, namely a title of the Post and a List of URL's of type String. It has the Pictures posted along with post.


```dart
// the FbPost Class for reference
class FbPost {

  String title;

  List<String> imagesUrl;

  FbPost({this.title, this.imagesUrl});
}
```

## Example

You can get the posts in an async method as shown below.

```dart

FacebookPost fbp = FacebookPost();
var myPost = [];
void getFbPosts() async {
    myPost = await fbp.getFbData('https://www.facebook.com/CBVpost/posts');
  }

```

The List so obtained could be shown as a ListView using ListTile as follows.

```dart
Widget myList(var theList) {
  return ListView.separated(
      separatorBuilder: (context, index) => Divider(
            thickness: 1,
          ),
      shrinkWrap: true,
      itemCount: theList.length,
      itemBuilder: (context, index) {
        print(theList[index].imagesUrl);
        var data = theList[index];
        return Padding(
            padding: const EdgeInsets.all(8.0),
            child: Container(
              color: Color(0xFFF0F0F0),
              child: ListTile(
                title: Text(data.title),
                subtitle: SizedBox(
                  height: 120,
                  width: 300,
                  child: ListView.separated(
                    separatorBuilder: (context, index) => VerticalDivider(
                      thickness: 1,
                    ),
                    scrollDirection: Axis.horizontal,
                    shrinkWrap: true,
                    itemCount: data.imagesUrl.length,
                    itemBuilder: (context, index) =>
                        Image.network(data.imagesUrl[index]),
                  ),
                ),
              ),
            ));
      });
}
```


CHANGELOG.md
## [0.0.1+1] - 2020-09-15
* Fixed README


## [0.0.1] - 2020-09-13

* Initial Publish. Added main files


test/facebook_posts_test.dart
// import 'package:flutter_test/flutter_test.dart';

// import 'package:facebook_posts/facebook_posts.dart';

// void main() {
//   test('adds one to input values', () {
//     final fb = FacebookPost();
//     expect(fb.getFbData(url), 3);
//     expect(calculator.addOne(-7), -6);
//     expect(calculator.addOne(0), 1);
//     expect(() => calculator.addOne(null), throwsNoSuchMethodError);
//   });
// }


lib/facebook_posts.dart
/// This package aims to get the data from Facebook Posts without accessing the
/// API of facebook.
/// Credits of this package go to ayob#4578 (discord).
///
/// This package also depends on two more packages. Namely the,
/// html and http packages.

import 'package:http/http.dart' as http;
import 'package:html/parser.dart';

/// The class Post defines the custom object used by the package to parse the
/// data obtained by the response fromt the website.

class FbPost {
  /// The title of the post is stored within the String variable named title
  String title;

  /// A list of strings to maintain the URL's of thumbnails of videos parsed
  /// from the given link of the facebook post;
  List<String> imagesUrl;

  FbPost({this.title, this.imagesUrl});
}

class FacebookPost {
  /// Main Helper function used to call the functionality of the package.
  Future getFbData(url) async {
    /// The url is parsed and passed to store and later aid in parsing the data obtained
    /// to acquire the title and the images.
    final response = await http.Client().get(Uri.parse(url));

    List<FbPost> posts = [];

    if (url.contains('facebook.com')) {
      // Check to see if the url is of facebook.com domain
      if (url.contains('posts')) {
        // Check to see if the url is of the proper format, including the posts word.
        try {
          if (response.statusCode == 200) {
            final document = parse(response.body);
            final length =
                document.getElementsByClassName('userContentWrapper').length;

            for (int i = 0; i <= length - 1; i++) {
              List<String> imageUrls = [];
              String title;
              final pinned = document
                  .getElementsByClassName('userContentWrapper')[i]
                  .children[0]
                  .children
                  .length;

              if (pinned != 4) {
                // To not show pinned publications
                try {
                  final images = document
                      .getElementsByClassName('userContentWrapper')[i]
                      .children[0]
                      .children[2]
                      .children[2]
                      .children[0]
                      .children[0]
                      .getElementsByTagName('img');
                  for (int index = 0; index <= images.length - 1; index++) {
                    String image = document
                        .getElementsByClassName('userContentWrapper')[i]
                        .children[0]
                        .children[2]
                        .children[2]
                        .children[0]
                        .children[0]
                        .getElementsByTagName('img')[index]
                        .outerHtml
                        .replaceAll('amp;', '')
                        .replaceAll(RegExp('.+[c][=]["]'), '')
                        .replaceAll(RegExp('["].+'), '');
                    imageUrls.add(image);
                  }
                  title = document
                      .getElementsByClassName('userContentWrapper')[i]
                      .getElementsByClassName('userContent')[0]
                      .text;
                } catch (e) {
                  imageUrls = [];
                  title = document
                      .getElementsByClassName('userContentWrapper')[i]
                      .getElementsByClassName('userContent')[0]
                      .text;
                }
              } else {
                imageUrls = null;
                title = null;
              }
              print('$i + $title + $imageUrls');
              if (imageUrls != null && title != null) {
                posts.add(FbPost(title: title, imagesUrl: imageUrls));
              }
            }
          }
        } catch (e) {
          print(e);
        }
      } else {
        print(
            'Check your URL format. Refer to the example.\nExample https://www.facebook.com/facebook/posts');
      }
    } else {
      print('The URL provided is not a valid facebook domain URL.');
    }
    return posts;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:facebook_posts/facebook_posts.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  FacebookPost fbp = FacebookPost();
  var myPost = [];

  void getFbPosts() async {
    myPost = await fbp.getFbData('https://www.facebook.com/CBVpost/posts');
  }

  Widget myWidget = Container();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        body: Center(
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          Container(
            height: MediaQuery.of(context).size.height * 0.4,
            child: Center(
              child: RaisedButton(
                child: Text('Press to get Post Data'),
                onPressed: () async {
                  await getFbPosts();
                  setState(() {
                    myWidget = myList(myPost);
                  });
                },
              ),
            ),
          ),
          Expanded(flex: 1, child: myWidget)
        ],
      ),
    ));
  }
}

Widget myList(var theList) {
  return ListView.separated(
      separatorBuilder: (context, index) => Divider(
            thickness: 1,
          ),
      shrinkWrap: true,
      itemCount: theList.length,
      itemBuilder: (context, index) {
        print(theList[index].imagesUrl);
        var data = theList[index];
        return Padding(
            padding: const EdgeInsets.all(8.0),
            child: Container(
              color: Color(0xFFF0F0F0),
              child: ListTile(
                title: Text(data.title),
                subtitle: SizedBox(
                  height: 120,
                  width: 300,
                  child: ListView.separated(
                    separatorBuilder: (context, index) => VerticalDivider(
                      thickness: 1,
                    ),
                    scrollDirection: Axis.horizontal,
                    shrinkWrap: true,
                    itemCount: data.imagesUrl.length,
                    itemBuilder: (context, index) =>
                        Image.network(data.imagesUrl[index]),
                  ),
                ),
              ),
            ));
      });
}
